File: src/main/java/xyz/nkomarn/composter/network/pipeline/PacketDecoder.java
package xyz.nkomarn.composter.network.pipeline;

import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.ReplayingDecoder;
import xyz.nkomarn.composter.network.Connection;
import xyz.nkomarn.composter.network.WrappedBuff;
import xyz.nkomarn.composter.network.protocol.packet.serverbound.ServerboundPacket;

import java.util.List;

import static xyz.nkomarn.composter.network.FlowDirection.SERVERBOUND;

public class PacketDecoder extends ReplayingDecoder<ServerboundPacket> {

    private final Connection connection;

    public PacketDecoder(Connection connection) {
        this.connection = connection;
    }

    @Override
    protected void decode(ChannelHandlerContext context, ByteBuf buff, List<Object> decoded) throws Exception {
        short packetId = buff.readUnsignedByte();
        var packet = (ServerboundPacket) connection.state().packetById(SERVERBOUND, packetId);

        if (packet == null) {
            return;
        }

        packet.decode(WrappedBuff.wrap(buff));
        decoded.add(packet);
    }
}

File: src/main/kotlin/kyta/composter/network/pipeline/PacketDecoder.kt
package kyta.composter.network.pipeline

import io.netty.buffer.ByteBuf
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.ReplayingDecoder
import kyta.composter.protocol.ServerboundPacket
import kyta.composter.network.Connection
import kyta.composter.protocol.FlowDirection
import kyta.composter.protocol.MinecraftPacketBuffer
import org.slf4j.Logger

class PacketDecoder(
    private val connection: Connection,
    private val logger: Logger,
) : ReplayingDecoder<ServerboundPacket>() {
    override fun decode(context: ChannelHandlerContext, buf: ByteBuf, output: MutableList<Any>) {
        val definitions = connection.state.definitions[FlowDirection.SERVERBOUND] ?: return
        val id = buf.readUnsignedByte().toInt()
        val packet = definitions.getSerializer(id)
            ?.deserialize(MinecraftPacketBuffer(buf))
            ?: return logger.warn("received packet #$id which does not have a registered serializer for state ${connection.state}, ignoring..")

        output.add(packet)
        logger.debug("received packet #$id (${packet::class.simpleName}) with state ${connection.state}")
    }
}